File: DSA/Sorting/Heap_sort.cpp
// C++ program for implementation of Heap Sort
#include <iostream>

using namespace std;


void heapify(int arr[], int n, int i)
{
	int largest = i; 
	int l = 2 * i + 1;
	int r = 2 * i + 2; 
	if (l < n && arr[l] > arr[largest])
		largest = l;
	if (r < n && arr[r] > arr[largest])
		largest = r;

	if (largest != i) {
		swap(arr[i], arr[largest]);
		heapify(arr, n, largest);
	}
}
void heapSort(int arr[], int n)
{
	for (int i = n / 2 - 1; i >= 0; i--)
		heapify(arr, n, i);
	for (int i = n - 1; i > 0; i--) {
		swap(arr[0], arr[i]);
		heapify(arr, i, 0);
	}
}

void printArray(int arr[], int n)
{
	for (int i = 0; i < n; ++i)
		cout << arr[i] << " ";
	cout << "\n";
}

int main()
{
    int n;
    cin>>n;
    int arr[n];
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
	heapSort(arr, n);

	cout << "Sorted array is \n";
	printArray(arr, n);
}

File: Python/heapsort.py
def heapify(arr, n, i):
   largest = i 
   l = 2 * i + 1
   r = 2 * i + 2 

   if l < n and arr[i] < arr[l]:
      largest = l

   if r < n and arr[largest] < arr[r]:
      largest = r

   if largest != i:
      arr[i],arr[largest] = arr[largest],arr[i] 
 
      heapify(arr, n, largest)

def heapSort(arr):
   n = len(arr)

   for i in range(n, -1, -1):
      heapify(arr, n, i)

   for i in range(n-1, 0, -1):
      arr[i], arr[0] = arr[0], arr[i] 

arr = [2,5,3,8,6,5,4,7]
heapSort(arr)
n = len(arr)
print ("Sorted array is")
for i in range(n):
   print (arr[i],end=" ")
